src/odbc_connection.cpp

```diff
@@ -2782,108 +2843,340 @@ fetch_and_store
+
+              // If the data is null, simply indicate and continue to the next
+              // column
+              if (string_length_or_indicator == SQL_NULL_DATA)
+              {
+                row[column_index].size = SQL_NULL_DATA;
+                continue;
+              }
+              // If the data (+ whatever null terminator, not included in
+              // string_length_or_indicator) was larger than the size of the
+              // buffer, then need to call SQLGetData at least once more
+              else if (
+                string_length_or_indicator == SQL_NO_TOTAL || 
+                string_length_or_indicator >= buffer_size
+              ) 
+              {
+                // Continue looping as long as we don't know the final resize
+                // that we need. Once a buffer size is returned instead of
+                // SQL_NO_TOTAL, we still need to call SQLGetData one more time,
+                // and the break out of this loop.
+                bool last_get_data_call = false;
+                while (!last_get_data_call)
                 {
-                  SQLCHAR *memory_start = (SQLCHAR *)data->bound_columns[column_index].buffer + (row_index * data->columns[column_index]->buffer_size);
-                  row[column_index].size = strlen((const char *)memory_start);
-                  // Although fields going from SQL_C_CHAR to Napi::String use
-                  // row[column_index].size, NUMERIC data uses atof() which requires
-                  // a null terminator. Need to add an aditional byte.
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size + 1]();
-                  memcpy
+                  // If not SQL_NO_TOTAL, we know how large to resize to buffer
+                  // to get the rest of the data.
+                  if (string_length_or_indicator != SQL_NO_TOTAL)
+                  {
+                    last_get_data_call = true;
+                  }
+
+                  // Handling the data that was returned depends heavily on the
+                  // target type of the binding data.
+                  switch(data->columns[column_index]->bind_type)
+                  {
+                    case SQL_C_BINARY:
+                      data_returned_length = buffer_size;
+                      buffer_size =
+                        string_length_or_indicator == SQL_NO_TOTAL ?
+                        buffer_size * 2 :
+                        row[column_index].size + string_length_or_indicator;
+                      row[column_index].char_data =
+                        (SQLCHAR *)
+                        realloc
+                        (
+                          row[column_index].char_data,
+                          buffer_size
+                        );
+                      row[column_index].size += data_returned_length;
+                      target_buffer =
+                        row[column_index].char_data + row[column_index].size;
```

Comment:
Is the size supposed to contain the current "max size" or the "data size". It seems to be used interchangeably here.

Reply:
The current "data size". The idea here is to point `target_buffer` (which is the pointer passed to `SQLGetData`) to the base `row[column_index].char_data` + an offset of how much _valid_ data there already exists there.

I think the underlying confusion is about `buffer_size`, which I am using in a couple different ways. I'm going to work on it a bit tonight, then maybe try again in the morning with fresh eyes.